app/controllers/application_controller.rb

```diff
@@ -4,30 +4,34 @@ class ApplicationController < ActionController::Base
 
 private
 
-  def session_data(id = estimate_id)
-    session[session_key(id)] ||= {}
+  def assessment_id
+    Digest::SHA256.hexdigest(assessment_code + (cookies["SessionData"] || ""))
   end
 
-  def session_key(id)
-    "estimate_#{id}"
+  def session_data
+    session[assessment_id] ||= {}
   end
 
-  def track_page_view(assessment_id: nil, page: page_name)
-    track("page_view", assessment_id:, page:)
+  def track_page_view(page: page_name)
+    track("page_view", page:)
   end
 
-  def track_validation_error(assessment_id: nil, page: page_name)
-    track("validation_message", assessment_id:, page:)
-    track_page_view(assessment_id:, page:)
+  def track_validation_error(page: page_name)
+    track("validation_message", page:)
+    track_page_view(page:)
   end
 
-  def track(event_type, assessment_id:, page:)
+  def track(event_type, page:)
     AnalyticsService.call(event_type:,
                           browser_id: cookies[BROWSER_ID_COOKIE],
-                          assessment_id:,
+                          assessment_code:,
                           page:)
   end
 
+  def assessment_code
+    nil
+  end
```

Comment:
This looks strange - should this method exist if it's being overridden by derived classes?

Reply:
It's not overriden by all derived classes, but it's _used_ by all of them (via `#track`). In cases where there is no assessment, but the track method want to be able to retrieve whatever the assessment code is, I think it's helpful having a `nil` default.